10-JavaScript/classes.md

```diff
@@ -0,0 +1,289 @@
+
+Fortunately as of [ES6](http://es6-features.org/#ClassDefinition), JavaScript has added some syntactical sugar to let us write classes and the JavaScript interpreter will convert that into __prototypical__ objects.  
+
+**So why cover this?**  It's important to know when you inspect and examine JavaScript code in debuggers that behind the scenes things are actually done using functions and prototypes.  Classes are simply syntax sugar to make our lives as devs easier.  Also as a developer, you will often work on an existing codebase that predates classes, and with team members who still write JavaScript code using the older methods.  
+
+## Introducing Classes
+
+In Ruby classes look like this:
+
+```ruby
+class Book
+  attr_reader :title, :author
+  def initialize(name, author)
+    @title = title
+    @author = author
+  end
+
+  def to_s      # Each dog instance is unique
+    "#{@title} by #{@author}"
+  end
+end
+
+# Test it out
+my_book = book.new ("Pragmatic Thinking & Learning", "Hunt")
+puts my_book.to_s   # Instance method
+
+```
+
+When you create a new book instance the `initialize` method runs and sets the instance variables to starting values.  
+
+Below is this Book class written in JavaScript.  It has a title, and author attributes and a `toString` method which outputs the object as a String.  Notice that everything in a class falls between the curly braces without any commas to separate each item.  
+
+```javascript
+class Book {
+  constructor(title, author) {
+    this.title = title;
+    this.author = author;
+  }
+
+  toString() {
+   return `${this.title} by ${this.author}`;
+  }
+}
+
+let warAndPeace = new Book("War and Peace", "Leo Tolstoy");
+
+console.log(warAndPeace.toString());
+> War and Peace by Leo Tolstoy
+```
+
+Classes can also be declared as an expression, similar to a function.
+
+```javascript
+const Book = class {
+ // code
+};
+```
+
+
+## Methods
+
+In a JavaScript class, methods are functions declared inside the class, without the `function` keyword.  There are a number of special functions such as __constructor__, __getter methods__ and __setter methods__.
+
+### Constructor Methods
+
+The constructor method is a special method named `constructor` which is called when a new instance is created.  You can think of it like an `initialize` method from Ruby.  Constructors exist to start our instances off in a proper state.  Use them to set your attributes with initial values using `this.<attributeName>` to create attributes (similar to using @ in Ruby).
+
+```javascript
+class Book {
+  constructor(title, author) {
+    console.log("Constructor Ran!");
+    this.title = title;
+    this.author = author;
+  }
+}
+
+let poodr = new Book("Practical Object Oriented Programming in Ruby", "Sandy Mietz");
+
+> Constructor Ran!
+```
+
+### Instance Methods
+
+Instance methods are created like a constructors with their own names.  
+
+```javascript
+const SALESTAX = 0.08;
```

Comment:
It seems like the expected syntax for adding a constant to the class in JS would be to define the class and then add `Book.SALESTAX = 0.08;` at the end rather than defining it outside of the class first.

Reply:
Except that then it wouldn't be a constant.  I kinda went back and forth on it and decided on just leaving it a constant at the top.